quote-book/synchronized/src/lib.rs

```diff
@@ -1,111 +1,392 @@
+    }
+
+    fn get_quotes(
+        &self,
+        pair: &Pair,
+        base_token_quantity: impl RangeBounds<u64>,
+        limit: usize,
+    ) -> Result<Vec<Quote>, QuoteBookError> {
+        self.synchronized_quote_book_impl
+            .get_quotes(pair, base_token_quantity, limit)
+    }
+
+    fn get_quote_ids(&self, pair: Option<&Pair>) -> Result<Vec<QuoteId>, QuoteBookError> {
+        self.synchronized_quote_book_impl.get_quote_ids(pair)
+    }
+
+    fn get_quote_by_id(&self, id: &QuoteId) -> Result<Option<Quote>, QuoteBookError> {
+        self.synchronized_quote_book_impl.get_quote_by_id(id)
+    }
+
+    fn num_scis(&self) -> Result<u64, QuoteBookError> {
+        self.synchronized_quote_book_impl.num_scis()
+    }
+}
+
+/// A callback for broadcasting a quote removal. It receives 1 argument:
+/// - A vector of quotes that have been removed
+pub type RemoveQuoteCallback = Arc<Mutex<dyn FnMut(Vec<Quote>) + Sync + Send>>;
+
+struct DbFetcherThread<DB: Ledger, Q: QuoteBook> {
+    db: DB,
+    quotebook: Q,
+    /// Message bus sender.
+    remove_quote_callback: RemoveQuoteCallback,
+    next_block_index: u64,
+    highest_processed_block_index: Arc<AtomicU64>,
+    stop_requested: Arc<AtomicBool>,
+    logger: Logger,
+}
+
+/// Background worker thread implementation that takes care of periodically
+/// polling data out of the database. Add join handle
+impl<DB: Ledger, Q: QuoteBook> DbFetcherThread<DB, Q> {
+    const POLLING_FREQUENCY: Duration = Duration::from_millis(10);
+    const ERROR_RETRY_FREQUENCY: Duration = Duration::from_millis(1000);
+
+    pub fn start(
+        db: DB,
+        quotebook: Q,
+        remove_quote_callback: RemoveQuoteCallback,
+        highest_processed_block_index: Arc<AtomicU64>,
+        stop_requested: Arc<AtomicBool>,
+        logger: Logger,
+    ) {
+        let next_block_index = highest_processed_block_index.load(Ordering::SeqCst);
+        let thread = Self {
+            db,
+            quotebook,
+            remove_quote_callback,
+            next_block_index,
+            highest_processed_block_index,
+            stop_requested,
+            logger,
+        };
+        thread.run();
+    }
+
+    fn run(mut self) {
+        log::info!(self.logger, "Db fetcher thread started.");
+        loop {
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Db fetcher thread stop requested.");
+                break;
+            }
+            let starting_block_index = self.next_block_index;
+            while self.load_block_data() && !self.stop_requested.load(Ordering::SeqCst) {}
+            if self.next_block_index != starting_block_index {
+                let last_processed_block_index = self.next_block_index - 1;
+                let mut quotes = self
+                    .quotebook
+                    .remove_quotes_by_tombstone_block(last_processed_block_index);
+                while quotes.is_err() {
```

Comment:
I think looping infinitely here is questionable, but separately, maybe we should move the retries into a `RetryingQuoteBook` that wraps a `QuoteBook` and add the retry behavior? (in a separate PR)

Reply:
I think I'll leave this for now. And I'll add an item to wrap the retryingquotebook separately.

Reply:
i think i suggested to loop infinitely -- it seems to me that, if we don't successfully remove things from the quotebook, then we may just end up with junk quotes in the book that are never removed?